### PR TITLE
feat: unify call participant search with @-mention matcher

### PR DESCRIPTION
**Pull request by XyneSpaces**
https://github.com/user-attachments/assets/8c742d7e-bc91-4566-a9ee-639589695d0a

## What
Unifies user/participant search across all three call modals (Instant call, Scheduled call, and the thread "add participants" modal) so they use the **same Fuse.js matcher that powers `@`-mention search**, replacing the naive `label.includes(query)` substring filter.

## Why
Users reported call-modal participant search was inferior to `@`-mention search:
- substring matching misses display-name vs old-name mismatches (e.g. typing `@V` should surface "Venkatesh"),
- prefix / word-boundary matches were not ranked first,
- deactivated users were not demoted.

The `@`-mention search already handles all of this via `searchUsers()` in `@xyne/shared/utils`. This PR reuses that matcher instead of maintaining a divergent substring path.

## Changes
- **New** `apps/dashboard/src/utils/participantSearch.ts` — `rankParticipantOptions(options, query)` splits options by `value` prefix and ranks users via `searchUsers()` and channels via `searchChannels()`; user groups pass through unchanged (already query-filtered upstream by `useUserGroupSearch`).
- `SearchParticipants.tsx` — new `disableClientFiltering` prop to skip the built-in substring filter when options are pre-ranked (exclusive-selection filtering still applies).
- `InstantCallModal.tsx` + `ScheduleCallModal.tsx` — wrap options with `rankParticipantOptions(...)` and set `disableClientFiltering`.
- `CallParticipantsSelectionModal.tsx` (thread add-participant) — rank channel members with `searchUsers()`; **non-members remain hidden (members-only)**, and pass `disableClientFiltering` to `EntityMultiSelector`.

## Scope / behavior
- Instant & Scheduled modals still search **all** workspace users + channels + groups (out-of-channel users allowed).
- Thread add-participant modal stays **members-only** — the `channelParticipantUserIds` restriction is retained, so non-channel users are not shown.
- Empty query preserves existing alphabetical ordering.

## Validation
- `tsc --noEmit` on the dashboard: no new errors in any touched file (pre-existing baseline errors are unrelated).
- Manual: search "venktesh" surfaces "V"; "harsha" surfaces both "Sai Harsha" and "Harshvardhan Bahukhandi"; "sk" ranks "Sk Sakil Mostak" first and tags out-of-channel; deactivated users sink.

## Notes
No ticket key was supplied for the branch name — happy to rename/retarget if a XYNE ticket exists.

### CODE REVIEW

File: apps/dashboard/src/utils/participantSearch.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  // Coerce name/email to strings before handing options to the matcher — some
+  // channel-member payloads arrive as `{ id, name }` with no email, and the
+  // rescoring step dereferences `.email`.
+  const rankedUsers =
+    users.length > 0
+      ? searchUsers(
+          users.map(u => ({
+            ...u,
+            name: typeof u.name === 'string' && u.name ? u.name : u.label,
+            email: typeof u.email === 'string' ? u.email : '',
```
Comment (XyneSpaces) [inline apps/dashboard/src/utils/participantSearch.ts]: **[major] R2** — `searchUsers` is configured with `minMatchCharLength: 2`, so single-character participant queries (for example the PR's `V`/`@V` case) will return no matches once client substring filtering is disabled. Add a one-character prefix/includes fallback here, or adjust the shared matcher before routing all non-empty queries through it.

File: apps/dashboard/src/components/Call/CallParticipantsSelectionModal/CallParticipantsSelectionModal.tsx [hunk 2]
```diff
@@ -74,15 +75,13 @@ export const InstantCallModal: React.FC<InstantCallModalProps> = ({
...
   const filteredUsers = useMemo(() => {
     if (!searchQuery.trim()) return channelUsers;
-    const query = searchQuery.toLowerCase();
-    return channelUsers.filter(user => {
-      const name = user.name?.toLowerCase() || '';
-      const email = user.email?.toLowerCase() || '';
-      return name.includes(query) || email.includes(query);
-    });
+    return searchUsers(channelUsers, searchQuery, channelUsers.length);
   }, [channelUsers, searchQuery]);
```
Comment (XyneSpaces) [inline apps/dashboard/src/components/Call/CallParticipantsSelectionModal/CallParticipantsSelectionModal.tsx:89]: **[major] R2** — this thread add-participant path calls `searchUsers` directly for every non-empty query; because the shared matcher uses `minMatchCharLength: 2`, a one-character search can show no channel members. Keep the old substring/prefix fallback for one-character queries or update `searchUsers` to support them.

File: apps/dashboard/src/utils/participantSearch.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  subtitle?: string | null;
+  name?: string;
+  email?: string;
+  displayName?: string | null;
+  status?: string | null;
+}
+
+export function rankParticipantOptions<T extends RankableOption>(options: T[], query: string): T[] {
+  // Empty query: preserve the caller's existing ordering (already alphabetised).
+  if (!query.trim()) return options;
```
Comment (XyneSpaces) [inline apps/dashboard/src/utils/participantSearch.ts:59]: **[major] R7** — this new shared ranking utility has no automated coverage for the tricky behavior it centralizes (one-character queries, fuzzy old-name/display-name matches, channel ranking, group passthrough, and deactivated-user demotion). Add unit tests around `rankParticipantOptions` so future changes don't silently regress the participant search paths.

**Review by XyneSpaces — COMMENTED** [review]
Inline findings from Xyne Spaces Reviewer.

**Comment by XyneSpaces** [comment]
## Xyne Spaces Reviewer — PR #328

**Verdict:** Changes requested
**Reviewed commit:** `8c8afbb795fa7b56b47ec6c54e8f4f83baaea433`
**Reviewed:** 5 files, 14 hunks · **Findings:** 0 blocking, 3 major, 0 nits

### Findings
| Severity | Rule | Location | Issue |
|----------|------|----------|-------|
| major | R2 | `apps/dashboard/src/utils/participantSearch.ts:57` | `rankParticipantOptions` routes all non-empty user searches through `searchUsers`, but that matcher has `minMatchCharLength: 2`, so one-character queries like `V`/`@V` can return no matches. |
| major | R2 | `apps/dashboard/src/components/Call/CallParticipantsSelectionModal/CallParticipantsSelectionModal.tsx:85` | The thread add-participant path calls `searchUsers` directly and has the same one-character search regression. |
| major | R7 | `apps/dashboard/src/utils/participantSearch.ts:36` | New shared participant ranking behavior has no automated tests for fuzzy/name/display-name/channel/group/deactivated edge cases. |

### Recommended reviewer
**Not assigned** — no CODEOWNERS file was found, GitHub path-commit ownership lookup hit rate limits, and the available GitHub tools do not expose a request-reviewers endpoint in this run. I did not invent a reviewer.

| Candidate | Owns (files in this PR) | Default reviewer? | Pending reviews | Tie-breaker |
|-----------|-------------------------|-------------------|-----------------|-------------|
| — | not computed: GitHub path commit lookup rate-limited | no CODEOWNERS found | not computed | no safe candidate identity to map to Spaces |

Reviewer request status: **not requested** due the tool/evidence failure above.

**Review by samitbarai — DISMISSED** [review]

File: apps/dashboard/src/utils/participantSearch.ts (hunk 3 above)
Comment (shubhamdhamija-juspay) [reply apps/dashboard/src/utils/participantSearch.ts:59]: No need for tests

File: apps/dashboard/src/components/Call/CallParticipantsSelectionModal/CallParticipantsSelectionModal.tsx (hunk 2 above)
Comment (shubhamdhamija-juspay) [reply apps/dashboard/src/components/Call/CallParticipantsSelectionModal/CallParticipantsSelectionModal.tsx:89]: Done

File: apps/dashboard/src/utils/participantSearch.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+});
+
+const toSearchableChannelOption = <T extends RankableOption>(option: T): T & { name: string } => ({
+  ...option,
+  name: typeof option.name === 'string' && option.name ? option.name : option.label,
+});
+
+export function rankParticipantOptions<T extends RankableOption>(options: T[], query: string): T[] {
+  // Empty query: preserve the caller's existing ordering (already alphabetised).
+  if (!query.trim()) return options;
```
Comment (harshpreet931) [inline apps/dashboard/src/utils/participantSearch.ts:59]: the 1 char guard should live here not in the caller searchUsers and searchChannels both set minMatchCharLength 2 so any single letter returns nothing and disableClientFiltering means there is no fallback left

File: apps/dashboard/src/components/Call/InstantCallModal/InstantCallModal.tsx [hunk 5]
```diff
@@ -131,6 +125,11 @@ export const InstantCallModal: React.FC<InstantCallModalProps> = ({
     );
   }, [activeUsers, channels, userGroups, allUsers, user?.id]);
 
+  const rankedParticipantOptions = useMemo(
```
Comment (harshpreet931) [inline apps/dashboard/src/components/Call/InstantCallModal/InstantCallModal.tsx:128]: this one still has the 1 char bug

File: apps/dashboard/src/components/Call/ScheduleCallModal/ScheduleCallModal.tsx [hunk 6]
```diff
@@ -508,6 +504,11 @@ export const ScheduleCallModal: React.FC<ScheduleCallModalProps> = ({
     fullUserList,
   ]);
 
+  const rankedParticipantOptions = useMemo(
```
Comment (harshpreet931) [inline apps/dashboard/src/components/Call/ScheduleCallModal/ScheduleCallModal.tsx:507]: same here no 1 char fallback

File: apps/dashboard/src/components/Call/CallParticipantsSelectionModal/CallParticipantsSelectionModal.tsx [hunk 7]
```diff
@@ -74,22 +75,23 @@ export const InstantCallModal: React.FC<InstantCallModalProps> = ({
...
+  // participant matcher. Some channel members have no email, which the raw
+  // `searchUsers` matcher assumes is always present.
+  const rankedChannelUsers = useMemo(() => {
+    const options = channelUsers.map(user => ({
+      ...user,
+      label: getUserDisplayName(user),
+      value: `user:${user.id}`,
+    }));
+    return rankParticipantOptions(options, searchQuery);
   }, [channelUsers, searchQuery]);
```
Comment (harshpreet931) [inline apps/dashboard/src/components/Call/CallParticipantsSelectionModal/CallParticipantsSelectionModal.tsx:89]: guard is correct but it is in the wrong layer it only fixes one of the three callers and duplicates logic the util should own can we drop it once rankParticipantOptions handles it

File: apps/dashboard/src/utils/participantSearch.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    if (opt.value.startsWith('user:')) users.push(opt);
+    else if (opt.value.startsWith('channel:')) channels.push(opt);
+    else groups.push(opt); // `user_group:` and any future prefixes
+  }
+
+  // Some channel-member payloads arrive as `{ id, name }` with no email, while
+  // the shared matcher dereferences both fields. Normalize them at this boundary.
+  const rankedUsers: T[] =
+    users.length > 0
+      ? searchUsers<SearchableUserOption<T>>(users.map(toSearchableUserOption), query, users.length)
```
Comment (harshpreet931) [inline apps/dashboard/src/utils/participantSearch.ts]: can we add a cap or a debounce

File: apps/dashboard/src/components/Call/CallParticipantsSelectionModal/CallParticipantsSelectionModal.tsx (hunk 7 above)
Comment (shubhamdhamija-juspay) [reply apps/dashboard/src/components/Call/CallParticipantsSelectionModal/CallParticipantsSelectionModal.tsx:89]: Moved to Partcipant search

File: apps/dashboard/src/components/Call/InstantCallModal/InstantCallModal.tsx (hunk 5 above)
Comment (shubhamdhamija-juspay) [reply apps/dashboard/src/components/Call/InstantCallModal/InstantCallModal.tsx:128]: fixed

File: apps/dashboard/src/components/Call/ScheduleCallModal/ScheduleCallModal.tsx (hunk 6 above)
Comment (shubhamdhamija-juspay) [reply apps/dashboard/src/components/Call/ScheduleCallModal/ScheduleCallModal.tsx:507]: Fixed

File: apps/dashboard/src/utils/participantSearch.ts (hunk 4 above)
Comment (shubhamdhamija-juspay) [reply apps/dashboard/src/utils/participantSearch.ts:59]: done

File: apps/dashboard/src/utils/participantSearch.ts (hunk 8 above)
Comment (shubhamdhamija-juspay) [reply apps/dashboard/src/utils/participantSearch.ts]: Done

**Review by harshpreet931 — APPROVED** [review]

**Review by samitbarai — APPROVED** [review]
